deployment/clouddeploy/deployer.go

```diff
@@ -561,32 +932,83 @@ func (d *Deployer) ModifyCluster(ctx context.Context, clusterID string, def *clu
 		return errors.Wrap(err, "failed to build cluster specs")
 	}
 
-	d.logger.Debug("generated new specification list", zap.Any("specs", newSpecs))
+	if !isServiceEqual(clusterInfo.Cluster.Services, newSpecs) {
+		d.logger.Info("cluster current spec is different from the def spec")
+		d.logger.Debug("generated new specification list", zap.Any("specs", newSpecs))
+		err = d.client.UpdateClusterSpecs(
+			ctx,
+			d.tenantID,
+			cloudProjectID,
+			cloudClusterID,
+			&capellacontrol.UpdateClusterSpecsRequest{
+				Specs: newSpecs,
+			})
+		if err != nil {
+			return errors.Wrap(err, "failed to update cluster specs")
+		}
 
-	err = d.client.UpdateClusterSpecs(
-		ctx,
-		d.tenantID,
-		cloudProjectID,
-		cloudClusterID,
-		&capellacontrol.UpdateClusterSpecsRequest{
-			Specs: newSpecs,
-		})
-	if err != nil {
-		return errors.Wrap(err, "failed to update cluster specs")
-	}
+		d.logger.Debug("waiting for cluster modification to begin")
+
+		err = d.mgr.WaitForClusterState(ctx, d.tenantID, cloudClusterID, "scaling")
+		if err != nil {
+			return errors.Wrap(err, "failed to wait for cluster modification to begin")
+		}
 
-	d.logger.Debug("waiting for cluster modification to begin")
+		d.logger.Debug("waiting for cluster to be healthy")
 
-	err = d.mgr.WaitForClusterState(ctx, d.tenantID, cloudClusterID, "scaling")
-	if err != nil {
-		return errors.Wrap(err, "failed to wait for cluster modification to begin")
+		err = d.mgr.WaitForClusterState(ctx, d.tenantID, cloudClusterID, "healthy")
+		if err != nil {
+			return errors.Wrap(err, "failed to wait for cluster to be healthy")
+		}
 	}
 
-	d.logger.Debug("waiting for cluster to be healthy")
+	// Server version can only be modified in cloud deployer
+	if def.Deployer == "cloud" {
+		var (
+			clusterVersion = ""
+			serverImage    = ""
+			releaseId      = ""
+		)
+		for _, nodeGroup := range def.NodeGroups {
+			if clusterVersion == "" {
+				clusterVersion = nodeGroup.Version
+				serverImage = nodeGroup.Cloud.ServerImage
+			} else {
+				if clusterVersion != nodeGroup.Version || serverImage != nodeGroup.Cloud.ServerImage {
+					return errors.New("all node groups must have the same version and image")
+				}
+			}
+		}
 
-	err = d.mgr.WaitForClusterState(ctx, d.tenantID, cloudClusterID, "healthy")
-	if err != nil {
-		return errors.Wrap(err, "failed to wait for cluster to be healthy")
+		if clusterVersion != clusterInfo.Cluster.Config.Version && serverImage != "" {
```

Comment:
Why do we check that the cluster version doesn't match here?  I think it's valid to upgrade between two different server images of the same overall clusterVersion, and I believe that the clusterVersion would have been updated already as part of the modification above?

Reply:
yes, we should be handling cases where we can update images on same server version, actually i faced a problem while implementing it.  For now we are updating only when server version are different. 

The problem is - 
How are we going to check if serverImage in def file is different from serverImage present on capella cluster. listAll Cluster api does not return serverImages of capella clusters.

I check with cloud team, there is this internal support api which need's some bearer token same like override token,
```
curl --location 'https://api.sbx-7.sandbox.nonprod-project-avengers.com/internal/support/clusters/535017be-275c-4d57-90c3-c654ed9f2f3e' \
--header 'Authorization: Bearer cb-645AD509-7639-4EE8-B58E-2DE04EFCE1CB'
```
which returns serverImage of a cluster, which we can use to check if images are different and then update accordingly. 

If you agree i can add this path with Bearer token support and use it.

Reply:
Where does the Bearer token come from, if its something that is trivial to get a hold of, we probably should add support for fetching it, since its definitely part of the 'definition of a cluster'.  A related question here is whether or not doing a normal version update 'removes' any custom serverImage that has been specified, or if it keeps it as-is.    We could also get some clarity from the cloud team on how these mechanisms are meant to fit together, and it might be trivial for them to make some minor modifications to the APIs they provide which could significantly simplify our implementation.

Reply:
I checked with cloud team, We can get the bearer toke from aws secret manager, and it is fixed for dev/sbx env, so we can hardcode as default or constant values. And for prod i have to check with SRE to get this token.

i am getting by normal version upgrade you meant updating server version without passing image, check whether previous server image is still present.  I tried to update server version without passing image, and it is failing to upgrade. so We need to specify image while updating server version, so serverImage will be changing each time we udpate server version. 

I think yes, i need to collect more info from cloud team on internal support api, to get serverImage. i will file a bug and work on it.

Reply:
I wonder if the override token is the same as the bearer token?  Both appear to work the same way (they are both in the AWS secret manager, and are both hard-coded for dev/sbx).  If it is a separate token, we should modify the init command to require both of them to be provided together (or not enable "private" image support at all).  Once we have those tokens, we can add better management of server images to the modify command.  We can probably do that in a separate enhancement PR rather than continuing to block this particular PR, since this one does provide MVP capabilities (it just requires a bit of extra thinking from the user right now).

Reply:
yes sure, i will create a ticket to track this work

Reply:
created a issue in cbdino: https://github.com/couchbaselabs/cbdinocluster/issues/50